### PR TITLE
Add `InList` support for timestamp type. (#3449)

### LINKED ISSUE #3449: Support timestamp data type in In_list node
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Ted-Jiang**
# Which issue does this PR close?

Closes #3449 .

 # Rationale for this change

After this change we support 
```
❯ \d test
+---------------+--------------+------------+-------------+-----------------------------+-------------+
| table_catalog | table_schema | table_name | column_name | data_type                   | is_nullable |
+---------------+--------------+------------+-------------+-----------------------------+-------------+
| datafusion    | public       | test       | typeId      | Int32                       | NO          |
| datafusion    | public       | test       | eventTime   | Timestamp(Nanosecond, None) | YES         |
+---------------+--------------+------------+-------------+-----------------------------+-------------+
2 rows in set. Query took 0.009 seconds.
❯  select * from test where "eventTime" in (CAST ('2014-01-01 23:00:01' AS TIMESTAMP));
+--------+---------------------+
| typeId | eventTime           |
+--------+---------------------+
| 1      | 2014-01-01 15:00:01 |
+--------+---------------------+
1 row in set. Query took 0.012 seconds.
```

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by Ted-Jiang** [comment]
@alamb @liukun4515  PTAL

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 1]
```diff
@@ -1659,4 +1754,105 @@ mod tests {
...
+                Some(1388588401000000000),
+                None,
+            )),
+            lit(ScalarValue::TimestampMicrosecond(None, None)),
+            lit(ScalarValue::TimestampMicrosecond(
+                Some(1388588401000000001),
+                None,
+            )),
+        ];
+        for v in 1388588401000000001..(OPTIMIZER_INSET_THRESHOLD + 4) {
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/in_list.rs:1784]: Does this loop have any values?

I would have expected the loop to go to a value higher than the starting `1388588401000000001`

```suggestion
        let start_ts = 1388588401000000001;
        for v in start_ts..(start_ts + OPTIMIZER_INSET_THRESHOLD + 4) {
```

**Review by alamb — COMMENTED** [review]
Looks good to me @Ted-Jiang. This is great. Thank you  -- I think there may be a bug in the test, but I also may be mistaken there.

File: datafusion/physical-expr/src/expressions/in_list.rs (hunk 1 above)
Comment (Ted-Jiang) [reply datafusion/physical-expr/src/expressions/in_list.rs]: Yes 😂

**Review by alamb — APPROVED** [review]
LGTM -- thanks @Ted-Jiang

**Pull request merged**